tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -432,49 +434,129 @@ impl StackGraphLanguage {
+    /// probably get a parse error.)
+    pub fn builder_into_stack_graph<'a>(
+        &'a self,
+        stack_graph: &'a mut StackGraph,
+        file: Handle<File>,
+        source: &'a str,
+    ) -> Builder<'a> {
+        Builder::new(self, stack_graph, file, source)
+    }
+}
+
+pub struct Builder<'a> {
+    sgl: &'a StackGraphLanguage,
+    stack_graph: &'a mut StackGraph,
+    file: Handle<File>,
+    source: &'a str,
+    graph: Graph<'a>,
+    remapped_local_ids: HashMap<usize, NodeID>,
+    injected_node_count: usize,
+    span_calculator: SpanCalculator<'a>,
+}
+
+impl<'a> Builder<'a> {
+    fn new(
+        sgl: &'a StackGraphLanguage,
+        stack_graph: &'a mut StackGraph,
         file: Handle<File>,
-        source: &str,
-        globals: &mut Variables,
+        source: &'a str,
+    ) -> Self {
+        let span_calculator = SpanCalculator::new(source);
+        Builder {
+            sgl,
+            stack_graph,
+            file,
+            source,
+            graph: Graph::new(),
+            remapped_local_ids: HashMap::new(),
+            injected_node_count: 0,
+            span_calculator,
+        }
+    }
+
+    /// Executes this builder.
+    pub fn build(
+        mut self,
+        mut globals: Variables<'a>,
         cancellation_flag: &dyn CancellationFlag,
     ) -> Result<(), LoadError> {
         let mut parser = Parser::new();
-        parser.set_language(self.language)?;
+        parser.set_language(self.sgl.language)?;
         unsafe { parser.set_cancellation_flag(cancellation_flag.flag()) };
-        let tree = parser.parse(source, None).ok_or(LoadError::ParseError)?;
-
+        let tree = parser
+            .parse(self.source, None)
+            .ok_or(LoadError::ParseError)?;
         let parse_errors = ParseError::into_all(tree);
         if parse_errors.errors().len() > 0 {
             return Err(LoadError::ParseErrors(parse_errors));
         }
         let tree = parse_errors.into_tree();
 
-        let mut graph = Graph::new();
+        let root_node = self.inject_node(NodeID::root());
         globals
-            .add(ROOT_NODE_VAR.into(), graph.add_graph_node().into())
+            .add(ROOT_NODE_VAR.into(), root_node.into())
             .expect("Failed to set ROOT_NODE");
+        let jump_to_scope_node = self.inject_node(NodeID::jump_to());
         globals
-            .add(JUMP_TO_SCOPE_NODE_VAR.into(), graph.add_graph_node().into())
+            .add(JUMP_TO_SCOPE_NODE_VAR.into(), jump_to_scope_node.into())
             .expect("Failed to set JUMP_TO_SCOPE_NODE");
+        let file_name = format!("{}", &self.stack_graph[self.file]);
         globals
-            .add(
-                FILE_PATH_VAR.into(),
-                format!("{}", &stack_graph[file]).into(),
-            )
+            .add(FILE_PATH_VAR.into(), file_name.into())
             .expect("Failed to set FILE_PATH");
```

Comment:
Now that we have a user-facing mechanism that subsumes this, I wonder if we should leave this completely to the user. One question would be how to allow users to do that for the CLI?

Reply:
I agree. Leaving this for another day. (The day that root and jumpt-to loose their special status, perhaps.)